KAFKA-5899: Added Connect metrics for connectors

rhauch opened the pull request:
This PR is the first of several subtasks for [KAFKA-2376](https://issues.apache.org/jira/browse/KAFKA-2376) to add metrics to Connect worker processes. See that issue and [KIP-196 for details](https://cwiki.apache.org/confluence/display/KAFKA/KIP-196%3A+Add+metrics+to+Kafka+Connect+framework).

This PR adds metrics for each connector using Kafka’s existing `Metrics` framework. This is the first of several changes to add several groups of metrics, this change starts by adding a very simple `ConnectMetrics` object that is owned by each worker and that makes it easy to define multiple groups of metrics, called `ConnectMetricGroup` objects. Each metric group maps to a JMX MBean, and each metric within the group maps to an MBean attribute.

Future PRs will build upon this simple pattern to add metrics for source and sink tasks, workers, and worker rebalances.

rhauch committed:
KAFKA-5899 Added Connect metrics for connectors

Added metrics for each connector using Kafka’s existing `Metrics` framework. Since this is the first of several changes to add several groups of metrics, this change starts by adding a very simple `ConnectMetrics` object that is owned by each worker and that makes it easy to define multiple groups of metrics, called `ConnectMetricGroup` objects. Each metric group maps to a JMX MBean, and each metric within the group maps to an MBean attribute.

Future PRs will build upon this simple pattern to add metrics for source and sink tasks, workers, and worker rebalances.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:
```diff
@@ -172,13 +181,25 @@ protected static ConfigDef baseConfigDef() {
+                        CommonClientConfigs.METRICS_RECORDING_LEVEL_DOC)
+                .define(METRIC_REPORTER_CLASSES_CONFIG, Type.LIST,
+                        "", Importance.LOW,
+                        CommonClientConfigs.METRIC_REPORTER_CLASSES_DOC);
```
These new worker configs were moved from `DistributedWorkerConfig` to `WorkerConfig`, so they still are inherited by `DistributedWorkerConfig`.

rhauch commented:
@ewencp, @hachikuji: I'd appreciate a review. As mentioned above, this is the first of several subtasks for [KAFKA-2376](https://issues.apache.org/jira/browse/KAFKA-2376).

rhauch commented:
@kkonstantine would appreciate a review.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+     * @param name the name; may not be null
+     * @return the validated name; never null
+     */
+    static String makeValidName(String name) {
```
nit: Seems that this is not the right place for this method. It's now between static and non-static field members. It's used in the constructor, but still, it should probably be placed lower in the class.

rhauch replied:
I thought the general Java convention is to put static fields and methods before member fields and variables? Nothing explicit is mentioned in the [Coding Guide](http://kafka.apache.org/coding-guide.html), but existing Kafka code is kind of all over the map. Some newer code seems to follow this convention while other code buries the methods in the middle of the non-static methods.

kkonstantine replied:
True there's nothing concrete. Google Java Style Guide mentions a logical order but nothing specific. There are some old Sun conventions: http://www.oracle.com/technetwork/java/codeconventions-141855.html

and then, if you look at Intelij's rules, under Code Style -> Java -> Arrangement you get a feel of some other conventions. 

But beyond guidelines and given the absence of clear consensus my point is that, besides static initializer blocks, I'd expect member fields, constructors, methods in that rough order.

rhauch replied:
Moved to bottom of class along with the other static method.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+    private final String workerId;
+    private final Map<String, MetricGroup> groupsByName = new HashMap<>();
+
+    public ConnectMetrics(String workerId, WorkerConfig config, Time time) {
```
missing javadoc?

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+        return name;
+    }
+
+    private final Logger log;
```
Is there a specific reason that this is not static?

rhauch replied:
Nope. Fixed.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+     * @param tagKeyValues pairs of tag name and values
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName, String... tagKeyValues) {
```
A bit surprising encoding of key value pair arguments as String var args. Of course I see it's all over elsewhere in other metrics in AK. I wonder if that's the standard way to write this metrics-specific code. In any case, too late now, but wanted to leave a comment.

rhauch replied:
It is a bit strange, but it does make it _much_ easier to use.

kkonstantine replied:
I think key-pairs of strings as list is brittle. But we inherited that.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+     * @param tagKeyValues    pairs of tag name and values
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName, boolean includeWorkerId, String... tagKeyValues) {
```
Allow me to begin by saying that I admit and like the simplicity of this pattern. 

Still, if this method is frequently used, and if the `gets` >> `creates` (are expected to be way more), then an alternative would be: 

```java
private final ConcurrentMap<String, MetricGroup> groupsByName = new ConcurrentHashMap<>();
...
public MetricGroup group(String groupName, boolean includeWorkerId, String... tagKeyValues) {
      MetricGroup group = groupsByName.get(groupName);
      if (group == null) {
          Map<String, String> tags = tags(includeWorkerId ? workerId : null, tagKeyValues);
          group = new MetricGroup(groupName, tags);
          groupsByName.putIfAbsent(groupName, group);
      }
      return group;
}
```
which is almost identical in terms of code. 

This pattern might waste a few objects on initialization, but otherwise it relieves the map from global synchronization on `gets`. If both `gets` and `creates` are expected to be just a few it doesn't worth it probably. Can't say because the use of `group` in this PR is not demonstrated.

rhauch replied:
Agree. BTW, can't wait until we can start using `computeIfAbsent()`, which would make this even easier:

    public MetricGroup group(String groupName, boolean includeWorkerId, String... tagKeyValues) {
        return groupsByName.computeIfAbsent(groupName, key->{
            Map<String, String> tags = tags(includeWorkerId ? workerId : null, tagKeyValues);
            return new MetricGroup(key, tags);
        });
    }

Thanks for the suggestion.

kkonstantine replied:
You bet.

kkonstantine replied:
Sorry, I'm not sure I understood the point of agreement. `computeIfAbsent` will still need a `ConcurrentMap`. I was saying that `synchronized` is going to do more harm than good here. And the use of concurrent map is practically fine. You agree with that, or leaving it the way it is?

rhauch replied:
Yes, `ConcurrentMap` is better and I've changed the PR to reflect this. My point with the `computeIfAbsent()` is that it will be able to be even simpler and more efficient with JDK 8 (the `MetricGroup` is only created if it is needed). Until then, I went with your suggestion.

kkonstantine replied:
Oh, I just noticed. Then `synchronized` is not needed anymore.

kkonstantine replied:
It was removed from the other versions of `group` but not from here.

rhauch replied:
Gotcha. It's fixed now.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+     * @return the map of tags that can be supplied to the {@link Metrics} methods; never null
+     */
+    static Map<String, String> tags(String workerId, String... keyValue) {
+        if ((keyValue.length % 2) != 0)
```
bit-nitpick :) 
```
// modulo 2 operation
if ((keyValue.length & 1) != 0)
```
most definitely optimized out by JIT because the divisor is known (`2`). Is it too hard to read? Sometimes it's good to train ppl who read the code by example. Can't be more nitpicking than that and if you want to keep uniformity with other versions of `getTags` elsewhere never mind. Not 100% sure either, leaving here to ask what you think in general for such optimizations that are JIT optimizable too.

rhauch replied:
Yeah, this is a case where I think it does becomes harder to read, especially because as you say it will be optimized.

kkonstantine replied:
Yeah, there's some didactic aspect to a few lines that are just a bit harder to read of course. (for instance if it was `var` instead of `2` things would be different). But I was on the edge too. Fine with leaving it.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -195,10 +209,91 @@ public Connector connector() {
+
+        private boolean asBoolean(MetricName metric) {
+            double value = this.metricGroup.metrics().metric(metric).value();
+            return value > 0.00001d || value < -0.000001d;
```
Radius doesn't seem to be equal on the positive and negative side. 
Should this be: 
```java
private static final double EPSILON = 0.00001d;
...
return value > EPSILON || value < -EPSILON;
```
instead? (also not sure which accuracy you prefer for epsilon `0.00001d` or `0.000001d`?)

rhauch replied:
Great catch. +1

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+     * @return the validated name; never null
+     */
+    static String makeValidName(String name) {
+        Objects.requireNonNull(name);
```
Do we want a `ConnectException` here instead? Not sure.

rhauch replied:
I'd say no, because `IllegalArgumentException` is the convention. Besides, it'd be private except I have unit tests for it, and where it's called actually does the null check prior to calling. Any null here means it's a coding error.

kkonstantine replied:
Sure, I was referring to future coding errors that will be revealed only during runtime. The current use is fine. I'm fine with NPE.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+        this.workerId = workerId != null ? makeValidName(workerId) : "worker-" + CONNECT_WORKER_ID_SEQUENCE.getAndIncrement();
+        this.time = time;
+
+        Map<String, String> metricsTags = new LinkedHashMap<>(); // currently empty
```
Is this a placeholder for future use? `MetricsConfig` is initialized with an empty `LinkedHashMap` within its constructor?

rhauch replied:
Well, sort of. It took me a long time to figure out how to pass tags, and I started out passing some global tags that I've since removed. Happy to remove it.

kkonstantine reviewed (commented):
Looks good and in compliance with other metrics definitions in AK. 

Left a few comments, some of which are optional.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,220 @@
+
+    /**
+     * Get the worker identifier.
+     * @return the worker ID; never null
```
empty line needed

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -537,6 +540,14 @@ public String workerId() {
 
+    /**
+     * Get the {@link ConnectMetrics} that uses Kafka Metrics and manages the JMX reporter.
+     * @return the Connect-specific metrics; never null
```
empty line needed

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectMetricsTest.java:
```diff
@@ -0,0 +1,140 @@
+    private ConnectMetrics metrics;
+
+    @Before
+    public void beforeEach() {
```
Traditionally called `setup`, no?

rhauch replied:
Bad habit, sorry. Fixed.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectMetricsTest.java:
```diff
@@ -0,0 +1,140 @@
+        metrics = new ConnectMetrics(null, new WorkerConfig(WorkerConfig.baseConfigDef(), DEFAULT_WORKER_CONFIG), new MockTime());
+    }
+
+    @Test
```
do we need a `tearDown` that calls `metrics.stop` ?

rhauch replied:
yup.

rhauch committed:
KAFKA-5899 Minor corrections per review

rhauch commented:
@kkonstantine, thanks for the review. I've addressed all of your concerns except a few that I address above.

rhauch committed:
KAFKA-5899 Relocated static helper method

rhauch commented:
@kkonstantine I think all concerns have been addressed. Thanks again!

rhauch committed:
KAFKA-5899 Made static variable final

rhauch committed:
KAFKA-5899 Corrected findbugs error

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,227 @@
+     * @param time     the time; may not be null
+     */
+    public ConnectMetrics(String workerId, WorkerConfig config, Time time) {
+        this.workerId = workerId != null ? makeValidName(workerId) : "worker-" + CONNECT_WORKER_ID_SEQUENCE.getAndIncrement();
```
Why would `workerId` ever be `null`?  And does having the `CONNECT_WORKER_ID_SEQUENCE` really help since all your workers would just have ID = 1? If this is just for tests, seems better to just require the ID to be passed in since we effectively require it for `Worker` and everything else.

rhauch replied:
It's definitely null in tests where each test uses a different workerId in the metrics and preventing JMX clashes. This seemed to reflect other metric tests, so I went with it.

ewencp replied:
hmm, i guess it seemed weird since that's not the intended usage and seems like a hack for the tests. not the end of the world, but seemed weird to me. which tests are you referring to so i can compare?

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,227 @@
+        }
+
+        Map<String, String> tags() {
+            return Collections.unmodifiableMap(tags);
```
you can just do the conversion to unmodifiable map one time in the constructor. it looks like at the moment this is only accessed in tests anyway.

rhauch replied:
ack

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,227 @@
+        if ((keyValue.length % 2) != 0)
+            throw new IllegalArgumentException("keyValue needs to be specified in pairs");
+        Map<String, String> tags = new HashMap<>();
+        if (workerId != null && !workerId.trim().isEmpty()) {
```
might want to rename `workerId` so it doesn't shadow the member field. something like `workerIdOpt` could work

rhauch replied:
actually, the method is static so it doesn't shadow. I can still rename it if you think it's really necessary.

ewencp replied:
ah, right. nah, that's fine. just when reviewing I had the thought that if we guaranteed non-`null`/non-empty in the constructor, this wouldn't be necessary. i realized that it was actually intentional, but easy to miss when reviewing here and not getting the same highlighting as an IDE

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java:
```diff
@@ -0,0 +1,227 @@
+        Objects.requireNonNull(name);
+        name = name.trim();
+        if (!name.isEmpty()) {
+            name = name.replaceAll("[:/\\\\*?,;\\[\\]\\\\=]+", "-");
```
I know the naming thing has bit us in the past, is this same approach used elsewhere and/or how was it decided on? Specifically, metric name constraints really shouldn't be JMX specific if that is the case here, despite the fact that the metrics is so obviously JMX-inspired.

I can easily find https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/metrics/KafkaMetricsGroup.scala#L46 but nothing else. Have we not had the same problems because metrics w/ topic names in them already  have constraints on the naming?

If I am remembering correctly, I think maybe both @gwenshap and @junrao were involved in some discussions, I think `-` vs `_` was a problem at some point? Maybe one of them could chime in here.

ewencp replied:
Another way of asking this is, should this actually happen in Connect or should it be happening in `JmxReporter`?

rhauch replied:
I guess a question I have is whether the metric names must be validated to satisfy all reporters. If so, then since the metrics are created before the reporters are, don't we need to ensure the metrics satisfy all reporters' requirements. That seems onerous.

Another option that may or may not be practical is to expect each reporter to accept metric names with whatever limitations we allow, and to deal with any conversions if that isn't the case.

I'm happy to move this into a more reusable place and/or to use existing conventions/rules.

ewencp replied:
to me it seems like we can't possibly know what the constraints of all reporters would be and they don't provide an interface for validation, so it should be up to them to figure out how to substitute. but i've also asked some other folks to maybe chime in here who may have better context on how we've handled this elsewhere.

junrao replied:
We did the dot to _ conversion for Yammer metric mostly because reporters like Graphite typically use dot to represent hierarchy and quite a few people are using the existing Graphite reporter that may be confused with dot.

Since Kafka metric is new, we could just let individual reporter deal with this issue, instead of changing the metric name directly.

rhauch replied:
Oh, I did forget this: the `workerId` is current set to:

    advertisedUrl.getHost() + ":" + advertisedUrl.getPort()

which means that at least `:` will be in one of the names that make up the group name for all Connect metric groups, and JMX names cannot have `:` characters. So right off the bat, we're using names that aren't valid in JMX.

rhauch replied:
Actually, I take that back. The workerId will be used in the tags for a few of the Connect metric groups. But, the connector name is user-supplied, and it _is_ in most of Connect's metric groups. And the only restriction we have on connector names is in distributed mode (may not contain `/` character). So it'd be pretty easy to break JMX naming conventions.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -138,11 +149,13 @@ private void pause() {
 
                 case STARTED:
                     connector.stop();
+                    metrics.recordPaused();
```
this falls through anyway, so this is just repeated

rhauch replied:
ack

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -138,11 +149,13 @@ private void pause() {
                 case INIT:
                     statusListener.onPause(connName);
                     this.state = State.STOPPED;
+                    metrics.recordPaused();
```
There is a state change in the `catch` block that needs a `metrics.recordFailed()`

rhauch replied:
I've redesigned this in #3911, but I'll move that into this PR. It changes the `ConnectorMetricsGroup` to implement and wrap the current `ConnectorStatus.Listener`. This way, the state changes for the supplied listener and the metrics group are handled the same way.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -195,10 +210,91 @@ public Connector connector() {
+        }
+
+        public void recordStopped() {
+            this.statusRunning.record(0);
```
huh, i think this was missed during the kip review because of the confusion over how these would be expressed? seems odd that you can't differentiate all the states

rhauch replied:
Yeah, this was missed because I used the `WorkerConnector.State` for the KIP, but once I got into the code I realized that Enum doesn't handle all of the states (including paused). Would appreciate advice as to whether I should add the remaining states and update the KIP, or implement per the KIP.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -195,10 +210,91 @@ public Connector connector() {
+
+        private boolean asBoolean(MetricName metric) {
+            double value = this.metricGroup.metrics().metric(metric).value();
+            return value > EPSILON || value < -EPSILON;
```
you shouldn't really need `EPSILON`. java guarantees IEEE 754 for `float` and `double` and 0 is guaranteed to be exactly representable such that `==` should work.

rhauch replied:
This also will be removed with the redesign.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java:
```diff
@@ -121,11 +131,15 @@ public void testStartupAndShutdown() {
-        WorkerConnector workerConnector = new WorkerConnector(CONNECTOR, connector, ctx, listener);
+        WorkerConnector workerConnector = new WorkerConnector(CONNECTOR, connector, ctx, metrics, listener);
 
+        assertStoppedMetric(workerConnector);
```
A different way of interpreting the enum `INIT` is `Initializing` rather than `Initialized`. Maybe that phrasing would make these tests clearer? Using `stopped` here seems confusing because it seems weird to go from stopped -> initialized.

ewencp reviewed (commented):
Mostly minor questions or changes requested, but the metrics question really probably needs some input from someone who remembers the discussions/issues better, presumably someone who worked on core + metrics.

rhauch committed:
KAFKA-5899 Additional changes and improvements

rhauch commented:
@ewencp, @kkonstantine: latest PR addresses all concerns, except for @ewencp's metric name validation question.

rhauch committed:
KAFKA-5899 Changed to always use a non-null worker ID in the Connect metrics

rhauch committed:
KAFKA-5899 Removed unnecessary synchronization

kkonstantine reviewed (approved):
My comments have been addressed. LGTM given that other reviewers will be happy too

rhauch commented:
@ewencp, @junrao: Regarding the metric name checks, I'd like to make one more change to make the Connect metric names even more restrictive to match the KIP, which (following the lead of the existing Kafka metrics documentation) specifies the connector names and task names will be `[-.\w]+`. To deal with the Graphite case @junrao mentioned above, the code would be a bit more restrictive and replace with a single `-` any sequence of 1 or more characters _other than_ `-` or a word character (e.g., `[a-zA-Z_0-9]`).

Right now, the `JmxReporter` does no filtering or conversion of metric names, since according to the Kafka metrics documentation most should be valid JMX names. Any change to the `JmxReporter` to do more would be a separate issue and PR.

rhauch committed:
KAFKA-5899 Make metric names even more restrictive than KIP-196 specifies

KIP-196 specifies that the connector and task names are used as tags within the metric group names, and that these tag values will consist of the characters `[.-\w]+`. This matches the existing Kafka metric documentation (for client ID, etc.), but the `.` character can cause problems in Graphite where `.` signals a hierarchical name.

This commit changes the `makeValidName` utility to replace with a single `-` character all sequences of 1 or more `-` and word characters (e.g., `[a-zA-Z_0-9]`). Note that `.` will also be replaced. This is more restrictive than the KIP, and thus should be valid.

rhauch commented:
Okay, I think I've addressed all concerns, including that the metric group and attribute names satisfy what's specified in [KIP-196](https://cwiki.apache.org/confluence/display/KAFKA/KIP-196%3A+Add+metrics+to+Kafka+Connect+framework), which itself was based on the patterns in the existing [Kafka metrics documentation](https://kafka.apache.org/documentation/#monitoring).

rhauch commented:
More specifically, here are just two examples from the existing Kafka metrics documentation:
* [common producer/consumer metrics](https://kafka.apache.org/documentation/#selector_monitoring) use `client-id=([-.\w]+)`
* [producer metrics](https://kafka.apache.org/documentation/#producer_monitoring) use `topic=([-.\w]+)`

ewencp commented:
Ok, so aside from the metric name mangling this already LGTM. We've posted about this and how it relates to KIP-190 on the mailing list to see if we can push the mangling down into `JmxReporter`so it only has to apply for that reporter. We'll commit this as is, and if that change is made we'll update accordingly.
